File: solutions/Python/subarray_sum_equals_k_prefix_sum.py
from typing import List


def subarray_sum_prefix_sum(nums: List[int], k: int) -> int:
    count = {0: 1}
    prefix = 0
    result = 0
    for num in nums:
        prefix += num
        result += count.get(prefix - k, 0)
        count[prefix] = count.get(prefix, 0) + 1
    return result

print(subarray_sum_prefix_sum([1, 1, 1], k=2))  # 2
print(subarray_sum_prefix_sum([1, 2, 3], k=3))  # 2

File: solutions/Rust/subarray_sum_equals_k_prefix_sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut count: HashMap<i32, i32> = HashMap::new();
        count.insert(0, 1);
        let mut prefix = 0;
        let mut result = 0;
        for num in nums {
            prefix += num;
            result += count.get(&(prefix - k)).copied().unwrap_or(0);
            *count.entry(prefix).or_insert(0) += 1;
        }
        result
    }
}
